subsys/shell/backends/shell_rpmsg.c

```diff
@@ -0,0 +1,207 @@
+
+SHELL_RPMSG_DEFINE(shell_transport_rpmsg);
+SHELL_DEFINE(shell_rpmsg, CONFIG_SHELL_PROMPT_RPMSG, &shell_transport_rpmsg,
+	     CONFIG_SHELL_BACKEND_RPMSG_LOG_MESSAGE_QUEUE_SIZE,
+	     CONFIG_SHELL_BACKEND_RPMSG_LOG_MESSAGE_QUEUE_TIMEOUT, SHELL_FLAG_OLF_CRLF);
+
+static int rpmsg_shell_cb(struct rpmsg_endpoint *ept, void *data,
+			  size_t len, uint32_t src, void *priv)
+{
+	const struct shell_transport *transport = (const struct shell_transport *)priv;
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	struct shell_rpmsg_rx rx;
+
+	if (len == 0) {
+		return RPMSG_ERR_NO_BUFF;
+	}
+
+	rx.data = data;
+	rx.len = len;
+	if (k_msgq_put(&sh_rpmsg->rx_q, &rx, K_NO_WAIT) != 0) {
+		return RPMSG_ERR_NO_MEM;
+	}
+
+	rpmsg_hold_rx_buffer(ept, data);
+	sh_rpmsg->shell_handler(SHELL_TRANSPORT_EVT_RX_RDY, sh_rpmsg->shell_context);
+
+	return RPMSG_SUCCESS;
+}
+
+static int uninit(const struct shell_transport *transport)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	rpmsg_destroy_ept(&sh_rpmsg->ept);
+	sh_rpmsg->ready = false;
+
+	return 0;
+}
+
+static int init(const struct shell_transport *transport,
+		const void *config,
+		shell_transport_handler_t evt_handler,
+		void *context)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	struct rpmsg_device *rdev;
+	int ret;
+
+	if (sh_rpmsg->ready) {
+		return -EALREADY;
+	}
+
+	if (config == NULL) {
+		return -EINVAL;
+	}
+	rdev = (struct rpmsg_device *)config;
+
+	k_msgq_init(&sh_rpmsg->rx_q, (char *)sh_rpmsg->rx_buf, sizeof(struct shell_rpmsg_rx),
+		    CONFIG_SHELL_RPMSG_MAX_RX);
+
+	ret = rpmsg_create_ept(&sh_rpmsg->ept, rdev, CONFIG_SHELL_RPMSG_SERVICE_NAME,
+			       CONFIG_SHELL_RPMSG_SRC_ADDR, CONFIG_SHELL_RPMSG_DST_ADDR,
+			       rpmsg_shell_cb, NULL);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sh_rpmsg->ept.priv = (void *)transport;
+
+	sh_rpmsg->shell_handler = evt_handler;
+	sh_rpmsg->shell_context = context;
+	sh_rpmsg->ready = true;
+
+	return 0;
+}
+
+static int enable(const struct shell_transport *transport, bool blocking)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+
+	ARG_UNUSED(blocking);
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int write(const struct shell_transport *transport,
+		 const void *data, size_t length, size_t *cnt)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	int ret;
+	uint32_t write_len;
+	void *tx_buf;
+
+	*cnt = 0;
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	tx_buf = rpmsg_get_tx_payload_buffer(&sh_rpmsg->ept, &write_len, false);
+	if (tx_buf == NULL) {
+		return -ENOBUFS;
+	}
+
+	write_len = MIN(length, write_len);
+	memcpy(tx_buf, data, write_len);
+
+	ret = rpmsg_send_nocopy(&sh_rpmsg->ept, tx_buf, write_len);
+	if (ret < 0) {
+		rpmsg_release_tx_buffer(&sh_rpmsg->ept, tx_buf);
+		*cnt = length;
+		return ret;
+	}
```

Comment:
What about using `rpmsg_send `or `rpmsg_trysend ` instead? These functions would do the job:
- get the TX buffer,
- copy in rpmsg buffer,
- release the buffer in case of error
- return the effective length in case of success